daemon/api_apps.go

```diff
@@ -209,20 +210,43 @@ func getLogs(c *Command, r *http.Request, user *auth.UserState) Response {
 		return AppNotFound("no matching services")
 	}
 
-	serviceNames := make([]string, len(appInfos))
-	for i, appInfo := range appInfos {
-		serviceNames[i] = appInfo.ServiceName()
+	c.d.overlord.State().Lock()
+	allQuotas, err := servicestate.AllQuotas(c.d.overlord.State())
+	if err != nil {
+		c.d.overlord.State().Unlock()
```

Comment:
Maybe create a small helper here to make the locking a bit easier? Something like:
```go
diff --git a/daemon/api_apps.go b/daemon/api_apps.go
index b8a1365f38..bd3d7d5faa 100644
--- a/daemon/api_apps.go
+++ b/daemon/api_apps.go
@@ -181,6 +181,32 @@ func splitAppName(s string) (snap, app string) {
        return s, ""
 }
 
+func getNamespaces(st *state.State, appInfos []*snap.AppInfo) (map[string][]string, error) {
+       st.Lock()
+       defer st.Unlock()
+       allQuotas, err := servicestate.AllQuotas(st)
+       if err != nil {
+               return nil, fmt.Errorf("cannot get service quotas: %v", err)
+       }
+
+       // group service names by their namespace so we can make only one
+       // journctl call per namespace.
+       namespaces := make(map[string][]string)
+       for _, appInfo := range appInfos {
+               opts, err := servicestate.SnapServiceOptions(st, appInfo.Snap.InstanceName(), allQuotas)
+               if err != nil {
+                       return nil, fmt.Errorf("cannot get service options for %q: %v", appInfo.Snap.InstanceName(), err)
+               }
+               namespace := ""
+               if opts.QuotaGroup != nil {
+                       namespace = opts.QuotaGroup.JournalNamespaceName()
+               }
+               namespaces[namespace] = append(namespaces[namespace], appInfo.ServiceName())
+       }
+
+       return namespaces, nil
+}
+
 func getLogs(c *Command, r *http.Request, user *auth.UserState) Response {
        query := r.URL.Query()
        n := 10
@@ -210,29 +237,10 @@ func getLogs(c *Command, r *http.Request, user *auth.UserState) Response {
                return AppNotFound("no matching services")
        }
 
-       c.d.overlord.State().Lock()
-       allQuotas, err := servicestate.AllQuotas(c.d.overlord.State())
+       namespaces, err := getNamespaces(c.d.overlord.State(), appInfos)
        if err != nil {
-               c.d.overlord.State().Unlock()
-               return InternalError("cannot get service quotas: %v", err)
-       }
-
-       // group service names by their namespace so we can make only one
-       // journctl call per namespace.
-       namespaces := make(map[string][]string)
-       for _, appInfo := range appInfos {
-               opts, err := servicestate.SnapServiceOptions(c.d.overlord.State(), appInfo.Snap.InstanceName(), allQuotas)
-               if err != nil {
-                       c.d.overlord.State().Unlock()
-                       return InternalError("cannot get service options for %q: %v", appInfo.Snap.InstanceName(), err)
-               }
-               namespace := ""
-               if opts.QuotaGroup != nil {
-                       namespace = opts.QuotaGroup.JournalNamespaceName()
-               }
-               namespaces[namespace] = append(namespaces[namespace], appInfo.ServiceName())
+               return InternalError(err.Error())
        }
-       c.d.overlord.State().Unlock()
 
        sysd := systemd.New(systemd.SystemMode, progress.Null)
        logs := make([]io.ReadCloser, 0, len(namespaces))
```

Reply:
Really nice suggestion, I've added this!